summingbird-core/src/main/scala/com/twitter/summingbird/planner/DagOptimizer.scala

```diff
@@ -291,7 +291,7 @@ trait DagOptimizer[P <: Platform[P]] {
       // TODO: we need to case class here to not lose the irreducible which may be named
       case ValueFlatMappedProducer(in, fn) =>
         // we know that (K, V) <: T due to the case match, but scala can't see it
-        def cast[K, V](p: Prod[(K, V)]): Prod[T] = p.asInstanceOf[Prod[T]]
+        def cast[K, V](p: Prod[(K, V)]): Prod[T] = IdentityKeyedProducer[P, K, V](p).asInstanceOf[Prod[T]]
```

Comment:
Okay. This code was actually the predecessor of the Dag optimizer code. I bet we could remove it now and replace the remove name nodes dag rule. I feel like that might be a cleaner solution since this code is actually making a false assumption (the strip name nodes code above).